tests/integration.rs

```diff
@@ -41,6 +43,91 @@ pub static ENARX_BACKEND: Lazy<Option<String>> = Lazy::new(|| match var("ENARX_B
     Err(VarError::NotPresent) => None,
 });
 
+pub struct PathLock(PathBuf);
+
+impl From<PathBuf> for PathLock {
+    fn from(path: PathBuf) -> Self {
+        Self(path)
+    }
+}
+
+impl PathLock {
+    /// Create a new lock within `OUT_DIR` in `CARGO_MANIFEST_DIR`.
+    pub fn new_in_out_dir(name: impl AsRef<Path>) -> Self {
+        Path::new(CRATE).join(OUT_DIR).join(name).into()
+    }
+
+    /// Acquire lock at associated path.
+    pub fn lock(&self) -> io::Result<impl Drop + '_> {
+        fs::File::options()
+            .write(true)
+            .create_new(true)
+            .open(&self.0)?;
+
+        struct Guard<'a>(&'a Path);
+        impl Drop for Guard<'_> {
+            fn drop(&mut self) {
+                fs::remove_file(self.0).unwrap_or_else(|e| {
+                    panic!(
+                        "failed to remove lock at `{}`: {e}",
+                        self.0.to_string_lossy()
+                    )
+                });
+            }
+        }
+        Ok(Guard(&self.0))
+    }
+
+    /// Wait until the lock is unlocked.
+    ///
+    /// Returns `true` if unlocked within `timeout` or `false` otherwise.
+    pub fn wait_unlocked(&self, timeout: Duration) -> bool {
+        let start = Instant::now();
+        let interval = min(Duration::from_millis(100), timeout / 10);
+        while start.elapsed() < timeout {
+            if !self.0.exists() {
+                return true;
+            }
+            thread::sleep(interval)
+        }
+        !self.0.exists()
+    }
+
+    /// Calls `f` if `pred` returns `true`.
+    ///
+    /// Returns:
+    /// - `Ok(Some(..))` if `f` was called
+    /// - `Ok(None)` if `pred` returned false
+    /// - `Err(e)` if either locking failed due to I/O error or a lock was not released within
+    /// `timeout`
+    pub fn do_if<T>(
+        &self,
+        pred: impl Fn() -> bool,
+        f: impl FnOnce() -> T,
+        timeout: Duration,
+    ) -> io::Result<Option<T>> {
+        if !pred() {
+            return Ok(None);
+        }
+        match self.lock() {
+            Ok(_) if !pred() => Ok(None), // another thread managed to lock first
+            Ok(_) => Ok(Some(f())),       // lock acquired and `pred()` is still `true`, call `f()`
+            Err(e) if e.kind() == io::ErrorKind::AlreadyExists => {
+                // lock has already been acquired by another thread, wait for it to be released
```

Comment:
We need to check for the condition again, since this is a general-purpose lock.
In general, the retry should not lock again, but since we're dealing with I/O here - better safe, than sorry.